libraries/platforms/dpi-verilator/bsg_manycore_platform.cpp

```diff
@@ -0,0 +1,553 @@
+        } while ((err == BSG_NONSYNTH_DPI_NOT_WINDOW  // Not in window
+                  || err == BSG_NONSYNTH_DPI_SUCCESS) // Still got a packet
+                 && (drains <= cap)); // Still haven't drained the FIFO capacity
+        
+        if(drains == cap){
+                manycore_pr_err(mc, "%s: Failed to drain fifo %s\n", __func__,
+                                hb_mc_fifo_rx_to_string(type));
+                return HB_MC_FAIL;
+        }
+
+        return HB_MC_SUCCESS;
+}
+
+static int hb_mc_platform_dpi_init(hb_mc_platform_t *platform, std::string hierarchy)
+{
+        svScope scope;
+        int credits = 0, err;
+        VerilatorWrapper *top = platform->top;
+
+        top->eval();
+
+        platform->dpi = new bsg_nonsynth_dpi::dpi_manycore<HB_MC_CONFIG_MAX>(hierarchy + ".mc_dpi");
+        platform->ctr = new bsg_nonsynth_dpi::dpi_cycle_counter<uint64_t>(hierarchy + ".ctr");
+
+        return HB_MC_SUCCESS;
+}
+
+static void hb_mc_platform_dpi_cleanup(hb_mc_platform_t *platform)
+{
+        bsg_nonsynth_dpi::dpi_manycore<HB_MC_CONFIG_MAX> *dpi;
+        bsg_nonsynth_dpi::dpi_cycle_counter<uint64_t> *ctr;
+
+        dpi = platform->dpi;
+        ctr = platform->ctr;
+
+        delete dpi;
+        delete ctr;
+
+        return;
+}
+
+// These track active manycore machine IDs, and top-level
+// instantiations.
+static std::set<hb_mc_manycore_id_t> active_ids;
+static std::map<hb_mc_manycore_id_t,VerilatorWrapper*> machines;
+
+/**
+ * Clean up the runtime platform
+ * @param[in] mc    A manycore to clean up
+ */
+void hb_mc_platform_cleanup(hb_mc_manycore_t *mc)
+{
+        hb_mc_platform_t *platform = reinterpret_cast<hb_mc_platform_t *>(mc->platform); 
+
+        hb_mc_tracer_cleanup(&(platform->tracer));
+
+        hb_mc_profiler_cleanup(&(platform->prof));
+
+        hb_mc_platform_dpi_cleanup(platform);
+
+        // Remove the key
+        auto key = active_ids.find(platform->id);
+        active_ids.erase(key);
+
+        platform->id = 0;
+
+        // Ideally, we would clean up each platform in
+        // platforms. However, we can't guarantee that someone won't
+        // call init again, so we'll live with the memory leak for
+        // now. It's a rare case, since few people call init/cleanup
+        // and then continue their program indefinitley.
+
+        return;
+}
+
+/**
+ * Initialize the runtime platform
+ * @param[in] mc    A manycore to initialize
+ * @param[in] id    ID which selects the physical hardware from which this manycore is configured
+ * @return HB_MC_FAIL if an error occured. HB_MC_SUCCESS otherwise.
+ */
+int hb_mc_platform_init(hb_mc_manycore_t *mc, hb_mc_manycore_id_t id)
+{
+
+        int r = HB_MC_FAIL, err;
+        
+        hb_mc_platform_t *platform = new hb_mc_platform_t;
+        std::string hierarchy = "TOP.manycore_tb_top";
+        hb_mc_idx_t x, y;
+        hb_mc_config_raw_t rd;
+
+        // Uncomment THIS and the statement in verilator_top.sv to enable tracing
```

Comment:
Maybe a #define to enable tracing?

Reply:
I thought about this -- especially after the last couple of days where Tommy and I have been knee-deep in vcd files.

Yes, you're right. That would be the way to make it available. 

My opinion is that I don't want people to use verilator for waveform debugging, unless they're an expert. The amount of weird odds and ends that occur without Z/X and the size of vcd files make it pretty unfriendly. I want to make users have to overcome a hurdle to use it. 

I'm open to changing it. I'm also open to creating a .debug binary build option (perhaps another PR) that build the waveform generating executable.

Reply:
sounds like another PR :)